service/history/shard/controller_impl.go

```diff
@@ -250,47 +210,46 @@ func (c *ControllerImpl) getOrCreateHistoryShardItem(shardID int32) (*historySha
 		return nil, err
 	}
 
-	if info.Identity() == c.GetHostInfo().Identity() {
-		shardItem, err := newHistoryShardsItem(
-			c.Resource,
-			shardID,
-			c.engineFactory,
-			c.config,
-		)
-		if err != nil {
-			return nil, err
-		}
-		c.historyShards[shardID] = shardItem
-		c.metricsScope.IncCounter(metrics.ShardItemCreatedCounter)
+	if info.Identity() != c.GetHostInfo().Identity() {
+		return nil, serviceerrors.NewShardOwnershipLost(c.GetHostInfo().Identity(), info.GetAddress())
+	}
 
-		shardItem.logger.Info("", tag.LifeCycleStarted, tag.ComponentShardItem)
-		return shardItem, nil
+	shard, err := newContext(
+		c.Resource,
+		shardID,
+		c.engineFactory,
+		c.config,
+		c.shardClosedCallback,
+	)
+	if err != nil {
+		return nil, err
 	}
+	shard.start()
+	c.historyShards[shardID] = shard
+	c.metricsScope.IncCounter(metrics.ShardContextCreatedCounter)
 
-	return nil, serviceerrors.NewShardOwnershipLost(c.GetHostInfo().Identity(), info.GetAddress())
+	shard.logger.Info("", tag.LifeCycleStarted, tag.ComponentShardContext)
+	return shard, nil
 }
 
-func (c *ControllerImpl) removeShard(shardID int32, expected *historyShardsItem) (*historyShardsItem, error) {
+func (c *ControllerImpl) removeShard(shardID int32, expected *ContextImpl) (*ContextImpl, int64, error) {
```

Comment:
since remove & delete are always used together, why not move the shard.Stop within this function?
one more thing, is returning the error if expected != actual shard useful?